src/dft/backends/mklcpu/commit.cpp

```diff
@@ -31,109 +31,128 @@
-public:
-    commit_derived_impl(sycl::queue queue, const detail::dft_values<prec, dom>& config_values)
-            : detail::commit_impl(queue, backend::mklcpu) {
-        DFT_ERROR status = DFT_NOTSET;
-        if (config_values.dimensions.size() == 1) {
-            status = DftiCreateDescriptor(&handle, get_precision(prec), get_domain(dom), 1,
-                                          config_values.dimensions[0]);
-        }
-        else {
-            status = DftiCreateDescriptor(&handle, get_precision(prec), get_domain(dom),
-                                          config_values.dimensions.size(),
-                                          config_values.dimensions.data());
-        }
-        if (status != DFTI_NO_ERROR) {
-            throw oneapi::mkl::exception("dft/backends/mklcpu", "commit",
-                                         "DftiCreateDescriptor failed");
-        }
-
-        set_value(handle, config_values);
+/// Commit impl class specialization for MKLCPU.
+template <dft::detail::precision prec, dft::detail::domain dom>
+class commit_derived_impl : public dft::detail::commit_impl {
+private:
+    static constexpr DFTI_CONFIG_VALUE mklcpu_prec = to_mklcpu(prec);
+    static constexpr DFTI_CONFIG_VALUE mklcpu_dom = to_mklcpu(dom);
+    using mklcpu_desc_t = DFTI_DESCRIPTOR_HANDLE;
 
-        status = DftiCommitDescriptor(handle);
+public:
+    commit_derived_impl(sycl::queue queue, const dft::detail::dft_values<prec, dom> config_values)
+            : dft::detail::commit_impl(queue, backend::mklcpu) {
+        DFT_ERROR status = DFTI_NOTSET;
+        sycl::buffer<DFT_ERROR, 1> status_buffer{ &status, sycl::range<1>{ 1 } };
+        sycl::buffer<mklcpu_desc_t, 1> handle_buffer{ &device_handle, sycl::range<1>{ 1 } };
+
+        queue.submit([&](sycl::handler& cgh) {
+            auto handle_obj =
+                handle_buffer.template get_access<sycl::access::mode::read_write>(cgh);
+            auto status_obj =
+                status_buffer.template get_access<sycl::access::mode::read_write>(cgh);
+
+            host_task<detail::kernel_name<mklcpu_desc_t>>(cgh, [=]() {
+                std::int64_t local_err = DFTI_MEMORY_ERROR;
+                if (config_values.dimensions.size() == 1)
+                    local_err = DftiCreateDescriptor(handle_obj.get_pointer(), mklcpu_prec,
+                                                     mklcpu_dom, 1, config_values.dimensions[0]);
+                else
+                    local_err = DftiCreateDescriptor(handle_obj.get_pointer(), mklcpu_prec,
+                                                     mklcpu_dom, config_values.dimensions.size(),
+                                                     config_values.dimensions.data());
+
+                set_value(*handle_obj.get_pointer(), config_values);
+                if (local_err != DFTI_NO_ERROR) {
+                    throw oneapi::mkl::exception("dft", "commit", "DftiCreateDescriptor failed");
+                }
```

Comment:
```suggestion
                if (local_err != DFTI_NO_ERROR) {
                    throw oneapi::mkl::exception("dft/backends/mklcpu", "commit", "DftiCreateDescriptor failed");
                }
                set_value(*handle_obj.get_pointer(), config_values);
```

Reply:
@FMarno closing this as these are not revelant anymore, old constructor is refactored.